nbconvert/exporters/templateexporter.py

```diff
@@ -217,9 +223,26 @@ def _load_template(self):
         # template by name with extension added, then try loading the template
         # as if the name is explicitly specified.
         template_file = self.template_file
-        self.log.debug("Attempting to load template %s", template_file)
-        self.log.debug("    template_path: %s", os.pathsep.join(self.template_path))
-        return self.environment.get_template(template_file)
+        
+        try:
+            self._log_template_loading(template_file)
+            return self.environment.get_template(template_file)
+        except TemplateNotFound:
+            if self.template_file.endswith(self.template_extension):
+                try: 
+                    template_file = self.template_file[:-len(self.template_extension)]
+                    self._log_template_loading(template_file)
+                    return self.environment.get_template(template_file)
+                except TemplateNotFound:
+                    raise TemplateNotFound(template_file)
+            else:
+                try:
+                    template_file = self.template_file + self.template_extension
+                    self._log_template_loading(template_file)
+                    return self.environment.get_template(template_file)
+                except TemplateNotFound:
+                    raise TemplateNotFound(template_file)
```

Comment:
One of the primary use cases that this broke was nbsphinx. That's fixed now, but it's not obvious to me what is different between 

```python
from jinja2 import DictLoader
my_loader = DictLoader({'my_template': "{%- extends 'rst.tpl' -%}"})
class MyExporter(nbconvert.RSTExporter):
    template_file = 'my_template'
exporter = MyExporter(extra_loaders=[my_loader])
```
and 
```python
loader = jinja2.DictLoader({'nbsphinx-rst': RST_TEMPLATE})
      super(Exporter, self).__init__(
          template_file='nbsphinx-rst', extra_loaders=[loader],          
          config= traitlets.config.Config(
                 {'HighlightMagicsPreprocessor': {'enabled': True}}),		                 
          filters={	
                 'markdown2rst': markdown2rst,
                 'get_empty_lines': _get_empty_lines,
                 'extract_toctree': _extract_toctree,
                 'get_output_type': _get_output_type,
                 'json_dumps': json.dumps,
             })		             
 		 
```

And extension tolerant loader seems to not be activated in that case; my first attempt at this was to add additional logic to ExtensionTolerantLoader to look for the case where the extension was present and could instead be removed, but despite the logic being correct (as evidence by logging) trying to loading from there didn't work and that failure seemed to be related to the `_load_template` function.